### PR TITLE
allow `TOptional(null)` and generate monomorphs from it

### PR DESCRIPTION
**Pull request by Simn**
This PR reinterprets `TOptional(null)` to introduce monomorphs, which allows encoding unknown types explicitly in macro-generated type-hints. Since the construct currently has no meaning and just errors when decoded, this should be a safe change.

I'm not introducing a new ComplexType constructor because that would unnecessarily break some macros. We can change this later if we decide to add a new constructor for another reason, in which case it would make no difference.

There is no accompanying syntax for this at the moment, and I'm not sure introducing one would be worth it.

### CODE REVIEW

**Comment by back2dos** [comment]
Uhm, could you put this in layman's terms? ^^

**Comment by Simn** [comment]
It allows defining type-hints "with holes" from macros, e.g. `String -> Int -> ?` or `Array<?>`. The holes are then filled by normal type-inference.

**Comment by back2dos** [comment]
Not sure the argument "it currently just errors, so we can define it as we want" really holds, even though I tend to use it myself. The thing is, if you get a `ComplexType` that is `null` and you wrap it within `TOptional` then you're creating a monomorph unintentionally and that can be the beginning of a wonderful inferred type error cascade.

If I can make a suggestion, it would be to rather use something like this:

```haxe
package haxe.macro;
@:coreType abstract Unknown<Const> {} //actually I would prefer this to go toplevel, but that risks breaking some code
```

The `Const` parameter could be used to identify different ids. How they would be chosen is a good question, but assuming that can be solved, you get a syntax that is quite similar to how monorphs are printed and also a representation that `toComplexType` could use for monomorphs rather than nullifying them.

**Comment by Simn** [comment]
I'm not sure how I feel about special types like that. In particular, I'll have to check if they are referential transparent with regards to typedefs. I also don't think the Const parameter makes much sense here. We don't want users to specify any IDs.
